### ISSUE TITLE
Bad advice given with: no-new-array

### ISSUE THREAD

**Issue by Jason3S** [sindresorhus/eslint-plugin-unicorn#2406]
The advice to use `Array.from({length});` instead of `new Array(length)` should be avoided. It is an order of magnitude slower than any other form of creating an array.

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/f3ca6a08a825966f6b240e765e263a88445171b5/docs/rules/no-new-array.md?plain=1#L33-L36

The reason for using `new Array(length)` is that you already have an idea what the length might be. It gives JavaScript the chance to set aside some space, so it can be filled in.

As can be seen from the perf tests below: `Array.from` is much slower than all the others.

Here is a set of tests on Node v20.15.0:
<img width="689" alt="image" src="https://github.com/user-attachments/assets/1112e8d0-ab0d-429c-8ed1-5e472623b7ae">

The ratios are about the same for v22.5.1:
<img width="683" alt="image" src="https://github.com/user-attachments/assets/6542a3ef-b989-47de-8a3d-7f6be1697e66">

`Array.from` which is `Array.from({length})` is between 10 to 20x slower than the other methods.

Growing the array organically is about half the speed of pre-allocating the size.

<details>

<summary>Perf test code.</summary>

```ts
    // knownWords is an array of 10,000 strings.

    test('for of', () => {
        const result: number[] = [];
        for (const a of knownWords) {
            result.push(a.length);
        }
        return result;
    });

    test('for i', () => {
        const result: number[] = [];
        const words = knownWords;
        const len = words.length;
        for (let i = 0; i < len; i++) {
            result.push(words[i].length);
        }
        return result;
    });

    test('for i r[i]=v', () => {
        const words = knownWords;
        const result: number[] = [];
        const len = words.length;
        for (let i = 0; i < len; i++) {
            result[i] = words[i].length;
        }
        return result;
    });

    test('for i Array.from(words)', () => {
        const words = knownWords;
        const result: number[] = Array.from(words) as unknown as number[];
        const len = words.length;
        for (let i = 0; i < len; i++) {
            result[i] = words[i].length;
        }
        return result;
    });

    test('for i Array.from', () => {
        const words = knownWords;
        const result: number[] = Array.from({ length: words.length });
        const len = words.length;
        for (let i = 0; i < len; i++) {
            result[i] = words[i].length;
        }
        return result;
    });

    test('for i Array(size)', () => {
        const words = knownWords;
        const result: number[] = new Array(words.length);
        const len = words.length;
        for (let i = 0; i < len; i++) {
            result[i] = words[i].length;
        }
        return result;
    });
```

</details>

My suggestion is to explicitly allow `new Array(number)`, to remove the poor performing suggestion, and caution against `Array.from({length})`.

Note: `Array.from(existingArray)` performs just fine. The issue seems to be only when the object is acting like an array, but is not an actual array.

**Comment by fisker** [comment]
```
"unicorn/no-new-array": "off",
```

**Comment by Jason3S** [comment]
> ```
> "unicorn/no-new-array": "off",
> ```

This issue is about the advice given and the auto fix, not how to turn it off.

I have already turned it off.

I just did not expect a linting rule to replace high performing code with extremely poor performing code. Especially when it is a recommended rule.

**Comment by Jason3S** [comment]
Note replacing:

```ts
const data = new Array(length);
```

with:

```ts
const data = [];
data.length = length;
```

Has the same performance characteristics and is 10x faster than `const data = Array.from({length});`
